### DISCUSSION TITLE
Particle coordinates

### DISCUSSION THREAD

**Discussion by asarnow** [nextpyp/.github#100] [Q&A]
I'd appreciate some help understanding the particle coordinate convention in nextPYP as they relate to IMOD/3dmod.

I have K3 data (5760x4092) and the tomogram is 3072 px in Z and padded up to a square with 6x binning. Thus the tomogram dimensions are 1024x512x1024. I did figure out that the coordinates are binned by 2x again and flipped in Z. However, I also found the X and Y coordinates are actually in [0,767] and not [0,1023]. How (or where in the code) is the 768 size determined?

My second point of confusion - is the coordinate the particle center or offset to the box corner? If it is the box corner, is the particle radius stored in the last column of the data["web"] array? That number seems to be the particle radius divided by the binning factor. If that is correct, is the particle detection radius meant to be in voxels rather than Å as per the web UI?

Thank you!

**Comment by abartesaghi** [comment]
@asarnow, I'm not entirely sure what are you trying to do, but here are some thoughts.

If your images are 5760x4092 and you select the option `Reshape tilt-images into squares`, they will be reshaped to 5760x5760 (unless you also select the option `Resize squares to closest multiple of 512` in which case the images will be resized to 6144x6144). If the `Binning factor for reconstruction` in the `Tomogram reconstruction` tab is set to 6, your tomograms will have dimensions 960x512x960 (or 1024x512x1024 if you selected resizing to the closest multiple of 512).

From a user's perspective, nextPYP uses the same convention for particle coordinates as IMOD. For example, you can manually pick particles from nextPYP generated tomograms using `imod tomogram.rec` and creating an IMOD model with your manually picked particles. If you save the IMOD model with an `.spk` extension, you can import these coordinates from the Pre-processing block using the particle detection method `import (path to particle coordiantes required)` (from the `Particle detection` tab) and specifying the location of the `.spk` files using the option `Import particle coordinates (*.spk)`.

All coordinates in nextPYP refer to the particle center (they are not offset from the box corner).

Also, the particle detection radius should always be specified in Å (not voxels).

The `data["web"]` structure is used internally by nextPYP to keep things in sync between the user interface and the processing backend and it should be transparent to users so you shouldn't need to worry about that.

**Reply by asarnow** [reply]
Thanks, Alberto. I want to use nextPYP to pick particles and export the coordinates (for training pickYOLO or other external SVA uses) without going through the whole refinement block.

I determined the coordinates (from `data["web"]`, `data["box"]`, or the coordinates .txt files from `<project>/<preprocessing>/train`, they're all the same) have a 768px domain instead of 960px or 1024px by making test picks in the corners of a tomogram. I also determined that if XYZ instead of XZY tomograms are needed, that using `clip rotx` to transform PYP tomograms, and then `imodtrans -Y` to backtransform external coordinates, will produce correct coordinates for PYP.

As far as converting coordinates, I think that now I just need to know how the coordinate limit (768) is arrived at, for using different tomogram sizes in the future.

Related - what particle list are imported particles added to? The particle count shown on the preprocessing block went up but none of the lists have the imported particles and there doesn't seem to be a way to display them.

**Reply by asarnow** [reply]
Also - if I try to run refinement with the imported coordinates I see the below error messages. It's not quite the same as the one Juha reported in #37, I guess because I have only imported coordinates. But it's also the case as mentioned above that I am not sure what particle list is being used. There was a list selected in the preprocessing block that imported the coordinates.

```
19	Traceback (most recent call last):
20	  File "/opt/pyp/bin/run/pyp", line 3877, in <module>
21	    csp_swarm(args.file, parameters, int(args.iter), args.skip, args.debug)
22	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
23	    return func(*args, **kwargs)
24	  File "/opt/pyp/bin/run/pyp", line 2395, in csp_swarm
25	    [allboxes, allparxs] = csp_extract_coordinates(
26	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
27	    return func(*args, **kwargs)
28	  File "/opt/pyp/src/pyp/inout/metadata/core.py", line 1869, in csp_extract_coordinates
29	    [allboxes, allparxs] = tomo_extract_coordinates(
30	  File "/opt/pyp/src/pyp/inout/metadata/core.py", line 2007, in tomo_extract_coordinates
31	    refinement = project_params.resolve_path(parameters["refine_parfile_tomo"])
32	KeyError: 'refine_parfile_tomo'
33	2024-03-28 16:15:11 [ERROR] /opt/pyp/bin/run/pyp:3891 | PYP (cspswarm) failed
34	2024-03-28 16:15:14 [INFO] /opt/pyp/bin/run/pyp:3232 | Job 1136968_1 (v0.6.4) launching on emnoded25.cluster.local using 16 task(s) and 4 GB of RAM
35	2024-03-28 16:15:14 [INFO] pyp/refine/csp/particle_cspt.py:57 | Total time elapsed (csp_local_merge): 00h 00m 00s
36	2024-03-28 16:15:14 [INFO] /opt/pyp/bin/run/pyp:3178 | Filesystem      Size  Used Avail Use% Mounted on
37	2024-03-28 16:15:14 [INFO] /opt/pyp/bin/run/pyp:3178 | /dev/nvme1n1p1  7.0T  214G  6.8T   4% /scratch/pyp
38	Traceback (most recent call last):
39	  File "/opt/pyp/bin/run/pyp", line 3957, in <module>
40	    particle_cspt.merge_movie_files_in_job_arr(
41	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
42	    return func(*args, **kwargs)
43	  File "/opt/pyp/src/pyp/refine/csp/particle_cspt.py", line 380, in merge_movie_files_in_job_arr
44	    with open(movie_file) as f:
45	FileNotFoundError: [Errno 2] No such file or directory: 'stacks.txt'
46	2024-03-28 16:15:14 [ERROR] /opt/pyp/bin/run/pyp:3978 | PYP (csp_local_merge) failed
```

**Comment by abartesaghi** [comment]
The `*.txt` coordinates in the `train/` folder (and in the internal `data` structure) should match the dimensions of the binned tomograms (either 960 or 1024), i.e., they should not be expressed in the 768 range. This shouldn't happen, but maybe the tomogram dimensions changed at some point and the metadata became out of date? If so, forcing a re-run of the pre-processing block may fix the problem.

When importing particle coordinates from outside nextPYP (or when using any of our particle picking methods), the resulting coordinates will be stored in the Auto Particles list. Note that this will only happen after the pre-processing block is executed (you need to do this even if you are just importing coordinates from outside).

The error message from the refinement block indicates that you didn't specify a value for the `Alignments from sub-volume averaging (*.txt)` option in the Refinement tab. Once you run the pre-processing block after specifying the list of imported coordinates, you will get a `*_volumes.txt` file you can use for this option. Once you do that, the refinement block should run successfully.
